### PR TITLE
DB-12314 Sqlshell init .rc, file for pre-setting sqlshell settings

### PR DESCRIPTION
**Pull request by martinrupp**
## Short Description
adding a .sqlshellrc that is executed first when running sqlshell

## Long Description
We search for a `.sqlshellrc` in `(CWD)/.sqlshellrc`, then `$HOME/.sqlshellrc`.
If one of those is found, we execute the commands in it. We don't run both files, CWD has precedence.

If sqlshell.sh is used with `-f script.sql` option, we first execute `.sqlshellrc`, then `script.sql`.

Note that when using `spliceengine/sqlshell.sh`, the CWD is `spliceengine/splice_machine`.
When using sqlshell standalone artifact, use `sqlshell/.sqlshellrc`.

## New i18n options
Some new options for i18n for configuring *client-side* conversion of date/time/timestamp to strings (note that this does not change SERVER-SIDE conversions):
```
set dateformat 'YYYY-MM-dd';
set timeformat 'HH:mm:ss';
set timestampformat 'yyyy-MM-dd T HH:mm:ss.SSS';
```

## How to test
When testing in spliceengine, create a `.sqlshellrc`. Otherwise, if created a sqlshell bundle, create a .sqlshellrc in same directory. To test, save this into the file
```
set dateformat dd.MM.YYYY
```

Then execute sqlshell and run `values current_date`, it should return date in dd.MM.YYYY format.

### CODE REVIEW

**Comment by martinrupp** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by cloudspliceci** [comment]
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/3285/

**Comment by cloudspliceci** [comment]
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/3289/

File: db-tools-i18n/src/main/java/com/splicemachine/db/iapi/tools/i18n/LocalizedResource.java [hunk 1]
```diff
@@ -362,13 +376,13 @@ else if (type == Types.TIMESTAMP ) {
 		}
 
 	public String getDateAsString(Date d){
-		if (!enableLocalized){
+		if (!customFormatDate){
```
Comment (msirek) [inline db-tools-i18n/src/main/java/com/splicemachine/db/iapi/tools/i18n/LocalizedResource.java:379]: enableLocalization can also be called in ij.java for the LOCALIZEDDISPLAY option, so perhaps we still need to check enableLocalized:

> if (!enableLocalized && !customFormatDate){

Same comment for getTimeAsString and  getTimestampAsString

File: db-tools-ij/src/main/java/com/splicemachine/db/impl/tools/ij/utilMain.java [hunk 2]
```diff
@@ -212,13 +230,33 @@ public Object invoke(Object proxy, Method method, Object[] args) throws Throwabl
...
 
+    /**
+     * load properties file `sqlshell.rc`.
+     * when running from spliceengine, this will be the file in splice_machine/sqlshell.rc (because CWD is there)
+     * otherwise, just the file next to sqlshell.sh.
+     * @return properties from sqlshell.rc. on error: null
+     */
+    public static Properties getProperties() {
+        Properties prop = new Properties();
+        String fileName = "sqlshell.rc";
```
Comment (msirek) [inline db-tools-ij/src/main/java/com/splicemachine/db/impl/tools/ij/utilMain.java:241]: Usually rc files start with a ".", e.g. .bashrc, .sqlshellrc
Should we do the same?

File: db-tools-i18n/src/main/java/com/splicemachine/db/iapi/tools/i18n/LocalizedResource.java (hunk 1 above)
Comment (martinrupp) [reply db-tools-i18n/src/main/java/com/splicemachine/db/iapi/tools/i18n/LocalizedResource.java:379]: when `enableLocalization(boolean mode)` is called it is calling `getInstance().init();`, which will set `customFormatDate` correctly, see this code in lines 168-178:

```
public void init (String encStr, String locStr, String msgF){
  // ...
	if (enableLocalized && locale != null){
		formatDecimal = (DecimalFormat)DecimalFormat.getInstance(locale);
		formatNumber = NumberFormat.getInstance(locale);
		formatDate = DateFormat.getDateInstance(DateFormat.LONG,locale);
		formatTime = DateFormat.getTimeInstance(DateFormat.LONG,locale);
		formatTimestamp = DateFormat.getDateTimeInstance(DateFormat.LONG,
				DateFormat.LONG, locale);
		customFormatDate = true;
		customFormatTime = true;
		customFormatTimestamp = true;
	}
   //...
}
```

but i forgot to reset to false.

Comment (martinrupp) [reply db-tools-i18n/src/main/java/com/splicemachine/db/iapi/tools/i18n/LocalizedResource.java:378]: done

File: db-tools-ij/src/main/java/com/splicemachine/db/impl/tools/ij/utilMain.java (hunk 2 above)
Comment (martinrupp) [reply db-tools-ij/src/main/java/com/splicemachine/db/impl/tools/ij/utilMain.java]: done, renamed to `.sqlshellrc`

**Review by msirek — APPROVED** [review]

**Comment by martinrupp** [comment]
if you have a .sqlshellrc in pwd -> use that
otherwise use $HOME/.sqlshellrc

**Comment by martinrupp** [comment]
changed the implementation. .sqlshellrc is now not a Properties files, but a set of commands that can be executed in the shell. that way, you don't have to remember two different syntaxes
also, you could now in a .sqlshellrc
- set schema ABC;
- connect 'XXX';
-  autocommit off;

**Review by msirek — APPROVED** [review]

**Comment by cloudspliceci** [comment]
TEST SUCCEEDED +1

File: db-tools-ij/src/main/java/com/splicemachine/db/impl/tools/ij/utilMain.java [hunk 3]
```diff
@@ -270,9 +273,28 @@ public void goStart(LocalizedInput[] in)
         for (int ictr = 0; ictr < commandGrabber.length; ictr++) {
             commandGrabber[ictr].reInit(in[ictr]);
         }
+
+        runSqlShellRc();
         runScriptGuts();
     }
 
+    private void runSqlShellRc() {
```
[the remark this answers is not in the document]
Comment (ascend1) [reply db-tools-ij/src/main/java/com/splicemachine/db/impl/tools/ij/utilMain.java:281]: Right, this seems to be the common behavior.

**Review by ascend1 — APPROVED** [review]

**Review by carolp-503 — APPROVED** [review]
+1

File: db-tools-ij/src/main/java/com/splicemachine/db/impl/tools/ij/utilMain.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (martinrupp) [reply db-tools-ij/src/main/java/com/splicemachine/db/impl/tools/ij/utilMain.java:281]: ok sure

[the remark this answers is not in the document]
Comment (martinrupp) [reply db-tools-ij/src/main/java/com/splicemachine/db/impl/tools/ij/utilMain.java:281]: done

[the remark this answers is not in the document]
Comment (martinrupp) [reply db-tools-ij/src/main/java/com/splicemachine/db/impl/tools/ij/utilMain.java:281]: having these two files now results in an output and execution like this:
```
* = current connection
[ executing commands from file '/Users/martinrupp/.sqlshellrc'
splice> MAXIMUMDISPLAYWIDTH 0;
ELAPSED TIME = 1 milliseconds
splice> prompt clock off;
ELAPSED TIME = 0 milliseconds
splice> 
[ executing commands from file '/Users/martinrupp/.sqlshellrc' done

[ executing commands from file '/Users/martinrupp/spliceengine/splice_machine/.sqlshellrc'
splice> MAXIMUMDISPLAYWIDTH 10;
ELAPSED TIME = 1 milliseconds
splice> set timestampformat 'yyyy-MM-dd T HH:mm:ss.SSS';
splice> 
[ executing commands from file '/Users/martinrupp/spliceengine/splice_machine/.sqlshellrc' done
```

**Comment by martinrupp** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by cloudspliceci** [comment]
TEST SUCCEEDED +1

**Pull request merged**